manager/integration/tests/common.py

```diff
@@ -150,17 +150,62 @@ def delete_and_wait_pod(api, pod_name):
         name=pod_name,
         namespace='default',
         body=k8sclient.V1DeleteOptions())
+    wait_delete_pod(api, pod_name)
+
+
+def delete_and_wait_statefulset(api, client, statefulset):
+    pod_data = get_statefulset_pod_info(api, statefulset)
+
+    apps_api = get_apps_api_client()
+    apps_api.delete_namespaced_stateful_set(
+        name=statefulset['metadata']['name'],
+        namespace='default', body=k8sclient.V1DeleteOptions())
+
     for i in range(DEFAULT_POD_TIMEOUT):
-        ret = api.list_namespaced_pod(namespace='default')
+        ret = apps_api.list_namespaced_stateful_set(namespace='default')
         found = False
         for item in ret.items:
-            if item.metadata.name == pod_name:
+            if item.metadata.name == \
+              statefulset['metadata']['name']:
                 found = True
                 break
         if not found:
             break
         time.sleep(DEFAULT_POD_INTERVAL)
     assert not found
+    # We need to generate the names for the PVCs on our own so we can
+    # delete them.
+    client = get_longhorn_api_client()
+    for pod in pod_data:
+        # Wait on Pods too, we apparently had timeout issues with them.
+        wait_delete_pod(api, pod['pod_name'])
+        api.delete_namespaced_persistent_volume_claim(
```

Comment:
Should you able to find PV name from PVC? `pvc.spec.volumeName`. Depends on `failed` is not a good way.

Reply:
PV name is obtained from PVC. PV name is referenced in the code as pod['pv_name'], which is a dictionary generated from get_statefulset_pod_info, which grabs the pv_name using get_volume_name, which accesses the volume name using pvc.spec.volumeName.

The code depends on Failed here not to check if it's our PV, but to see if the PV (that we already know the name of) was created manually (any volumes created through a StorageClass won't fail like that) so we can clean it up properly. There are a few other ways I can think of to indicate that the PV was created manually, such as adding some kind of annotation or something. What do you think?

Reply:
I see now. Yeah, make sense.